ext/standard/array.c

```diff
@@ -893,17 +932,47 @@ static int php_array_user_compare(Bucket *f, Bucket *s) /* {{{ */
 	BG(user_compare_fci).params = args;
 	BG(user_compare_fci).retval = &retval;
 	BG(user_compare_fci).no_separation = 0;
-	if (zend_call_function(&BG(user_compare_fci), &BG(user_compare_fci_cache)) == SUCCESS && Z_TYPE(retval) != IS_UNDEF) {
-		zend_long ret = zval_get_long(&retval);
-		zval_ptr_dtor(&retval);
-		zval_ptr_dtor(&args[1]);
-		zval_ptr_dtor(&args[0]);
-		return ZEND_NORMALIZE_BOOL(ret);
-	} else {
-		zval_ptr_dtor(&args[1]);
-		zval_ptr_dtor(&args[0]);
+	call_failed = zend_call_function(&BG(user_compare_fci), &BG(user_compare_fci_cache)) == FAILURE || Z_TYPE(retval) == IS_UNDEF;
+	zval_ptr_dtor(&args[1]);
+	zval_ptr_dtor(&args[0]);
+	if (UNEXPECTED(call_failed)) {
 		return 0;
 	}
+
+	if (UNEXPECTED(Z_TYPE(retval) == IS_FALSE || Z_TYPE(retval) == IS_TRUE)) {
+		if (!ARRAYG(compare_deprecation_thrown)) {
+			php_error_docref(NULL, E_DEPRECATED,
+				"Returning bool from comparison function is deprecated, "
+				"return an integer less than, equal to, or greater than zero");
```

Comment:
> an integer less than, equal to, or greater than zero

Isn't that a little redundant? (as it basically includes all integers). Alternatively we could say "-1, 0 or 1".

Reply:
Yes, this is intentionally redundant to clarify meaning of the value. I used "-1, 0 or 1" here initially, but as @hikari-no-yume pointed out on the mailing list, there is no actual requirement to return one of those here, and some functions like strcmp() which are reasonable to use in this context don't return -1, 0, 1 specifically.

Reply:
Ah I see. 👍

Reply:
There could be a better way to word it, but there is a risk of the message being too long…